packages/node_modules/@ciscospark/media-engine-webrtc/test/unit/spec/engine.js

```diff
@@ -0,0 +1,776 @@
+              expectToRenegotiateOnTransition = expectToRenegotiateOnTransition || audioStart.includes(`send`) !== audioEnd.includes(`send`) && (audioStart.includes(`send`) || audioEnd.includes(`send`));
+              expectToRenegotiateOnTransition = expectToRenegotiateOnTransition || videoStart.includes(`send`) !== videoEnd.includes(`send`) && (videoStart.includes(`send`) || videoEnd.includes(`send`));
+
+              let sendingAudioChange = false;
+              sendingAudioChange = audioStart.includes(`send`) !== audioEnd.includes(`send`);
+
+              let sendingVideoChange = false;
+              sendingVideoChange = videoStart.includes(`send`) !== videoEnd.includes(`send`);
+
+              let receivingAudioChange = false;
+              receivingAudioChange = audioStart.includes(`recv`) !== audioEnd.includes(`recv`);
+
+              let receivingVideoChange = false;
+              receivingVideoChange = videoStart.includes(`recv`) !== videoEnd.includes(`recv`);
+
+              const expectNewRemoteTrackOnReturn = false;
+              let expectToRenegotiateOnReturn = expectNewRemoteTrackOnReturn;
+              expectToRenegotiateOnReturn = expectToRenegotiateOnReturn || audioStart.includes(`send`) !== audioEnd.includes(`send`) && (audioStart.includes(`send`) || audioEnd.includes(`send`));
+              expectToRenegotiateOnReturn = expectToRenegotiateOnReturn || videoStart.includes(`send`) !== videoEnd.includes(`send`) && (videoStart.includes(`send`) || videoEnd.includes(`send`));
+
+              console.log(`AAAAAAAA`, expectToRenegotiateOnReturn, expectNewRemoteTrackOnReturn, audioStart.includes(`send`), audioEnd.includes(`send`), videoStart.includes(`send`), videoEnd.includes(`send`));
+              console.log(`BBBBBBBB`, audioStart.includes(`send`), audioEnd.includes(`send`), videoStart.includes(`send`), videoEnd.includes(`send`));
+
+              const message = `when ${[
+                audioStart === audioEnd ? `audio is '${audioStart}'` : `audio changes from '${audioStart}' to '${audioEnd}'`,
+                videoStart === videoEnd ? `video is '${videoStart}'` : `video changes from '${videoStart}' to '${videoEnd}'`
+              ]
+                .filter((m) => m)
+                .join(` and `)}`;
+
+              describe(message, () => {
+                // Note: This describe block is stateful; `step()` statements must
+                // execute in order.
+                let engine;
+                before(() => {
+                  engine = new WebRTCMediaEngine();
+                  engine.on(`change:receivingVideo`, (value, previous) => console.log(`XXXXX receivingVideo`, engine.receivingVideo, value, previous));
+                });
+
+                after(() => {
+                  engine.stop();
+                });
+
+                step(`initiates a session with audio=${audioStart} and video=${videoStart}`, () => {
```

Comment:
`step`! Nice!

Reply:
@moriahmaney said that step and retries don't play nice, may need to confirm this one

Reply:
yea, i could see that, but with the ways these are staged, retries wouldn't help anyway.